src/compiler/checker.ts

```diff
@@ -5351,17 +5355,40 @@ module ts {
                 }
                 // Target type is type of prototype property
                 let prototypeProperty = getPropertyOfType(rightType, "prototype");
-                if (!prototypeProperty) {
-                    return type;
+                if (prototypeProperty) {
+                    let targetType = getTypeOfSymbol(prototypeProperty);
+                    if (targetType !== anyType) {
+                        // Narrow to the target type if it's a subtype of the current type
+                        if (isTypeSubtypeOf(targetType, type)) {
+                            return targetType;
+                        }
+                        // If the current type is a union type, remove all constituents that aren't subtypes of the target.
+                        if (type.flags & TypeFlags.Union) {
+                            return getUnionTypeOfSubtypeConstituents(<UnionType>type, targetType);
+                        }
+                    }
                 }
-                let targetType = getTypeOfSymbol(prototypeProperty);
-                // Narrow to target type if it is a subtype of current type
-                if (isTypeSubtypeOf(targetType, type)) {
-                    return targetType;
+                // Target type is type of constructor signiture
+                let constructSignatures: Signature[];
+                if (rightType.flags & TypeFlags.Interface) {
+                    constructSignatures = (<InterfaceTypeWithDeclaredMembers>rightType).declaredConstructSignatures;
                 }
-                // If current type is a union type, remove all constituents that aren't subtypes of target type
-                if (type.flags & TypeFlags.Union) {
-                    return getUnionType(filter((<UnionType>type).types, t => isTypeSubtypeOf(t, targetType)));
+                else if (rightType.flags & TypeFlags.Anonymous) {
+                    constructSignatures = (<ResolvedType>rightType).constructSignatures;

```

Comment:
Actually you can just call getSignaturesOfType once, without checking the type flags.